Lib/pathlib.py

```diff
@@ -583,6 +583,8 @@ def __len__(self):
             return len(self._parts)
 
     def __getitem__(self, idx):
```

Comment:
By the way, another option to consider (and this is a bit unusual so I will defer to others if there's an objection) would be to use `@functools.singledispatchmethod` for handling the slices in what may be a neater way:

```python
@functools.singledispatchmethod
def __getitem__(self, idx):
    if idx < 0 or idx >= len(self):
        raise Indexerror(idx)
    return self._pathcls._from_parsed_parts(self._drv, self._root,
                                            self._parts[:-idx - 1])

@__getitem__.register(slice)
def _(self, idx):
    return tuple(self[i] for i in range(*idx.indices(len(self))))
```

I tend to like single dispatch rather than multiple returns, but it might be weird in a dunder method. Plus, I can find no other uses of `@functools.singledispatchmethod` anywhere else in the standard library, so maybe you won't want to rock the boat in that regard 😛.

Reply:
I like that better for sure, but I'll wait to see if others have opinions.

Reply:
FTR, using `singledispatch` for this looks awkward to me.
(but perhaps pattern matching would help... ;-))

Reply:
Also, I think there's some precedence:
https://github.com/python/cpython/blob/8d59eb1b66c51b2b918da9881c57d07d08df43b7/Lib/tracemalloc.py#L204
https://github.com/python/cpython/blob/9f55551f3df238e58315e724e50cb0d574d75b94/Lib/sre_parse.py#L165
https://github.com/python/cpython/blob/85339f5c220a5e79c47c3a33c93f1dca5c59c52e/Lib/calendar.py#L60

Reply:
`functools.singledispatchmethod` was introduced in Python 3.8 I think. Those other implementations are much older and wouldn't have had this option (they may not have even had the option of `functools.singledispatch`).

If Antoine doesn't like it I'm fine with leaving it with a simple `isinstance` check (though hopefully this doesn't get used as precedent for "we never do that in CPython", since I think everyone involved is pretty ambivalent or apathetic about it 😛).